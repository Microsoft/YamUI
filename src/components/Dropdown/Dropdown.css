/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
@import 'src/css/variables/colors.css';
@import 'src/css/variables/fonts.css';

/* Label element above the Dropdown */
.y-dropdown .ms-Dropdown-label {
  @extend text__mediumSub;
  font-weight: $fontWeight__bold;
  color: $textColor__secondary;
  padding-bottom: 0.4rem;
  margin-bottom: 0;
}

/* Chevron Icon */
.y-dropdown .ms-Dropdown-caretDownWrapper {
  right: 1.1rem;
  height: 3.2rem;
  line-height: 3.4rem;
}
.y-dropdown .y-dropdown__caretDown {
  color: $linkColor;
}

/* Dropdown options */
.y-dropdown .ms-Dropdown-items {
  padding-bottom: 0.4rem;
}

.y-dropdown .ms-Dropdown-item:hover,
.y-dropdown .ms-Dropdown-item:active,
.y-dropdown .ms-Dropdown-item:focus {
  background-color: $interactiveElementBackround__focused;
}
.y-dropdown .ms-Dropdown-item[aria-selected='true'] {
  background-color: $interactiveElementBackround__selected;
}

/* Nested option content */
.y-dropdown--item {
  color: $textColor__secondary;
}

/* Section headers and dividers */
.y-dropdown .ms-Dropdown-header {
  padding: 0.7rem 1.2rem 0.9rem;
  margin: 0;
}
.y-dropdown--header {
  color: $textColor;
}
.y-dropdown .ms-Dropdown-divider {
  background-color: $borderColor__secondary;
  margin-bottom: -1px;
}

.y-dropdown--option-icon {
  padding-top: 0.1rem;
}

/* The selected option in the visible field area */
.y-dropdown .ms-Dropdown-title {
  color: $textColor;
  border-color: $borderColor;
  border-radius: 1px;
  height: 3.6rem;
  padding: 0.7rem 3rem 0.7rem 1.1rem;
}
.y-dropdown .ms-Dropdown-titleIsPlaceHolder {
  color: $textColor__metadata;
}
.y-dropdown .ms-Dropdown:active .ms-Dropdown-title,
.y-dropdown .ms-Dropdown:hover .ms-Dropdown-title,
.y-dropdown .ms-Dropdown:focus .ms-Dropdown-title {
  border-color: $borderColor__focused;
}

.y-dropdown .ms-Dropdown-title .y-dropdown--ellipsed-title {
  @extend ellipsis;
}

/* Non-layered callout styling */
.y-dropdown .ms-Callout-container {
  z-index: 1000;
}
