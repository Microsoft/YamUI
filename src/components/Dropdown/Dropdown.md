### Notes for use

A `Dropdown` can be used in a <a href="https://facebook.github.io/react/docs/forms.html#controlled-components" target="_blank">controlled</a> or <a href="https://facebook.github.io/react/docs/uncontrolled-components.html" target="_blank">uncontrolled</a> manner.

### Examples

Controlled dropdown:

```js { "props": { "data-description": "controlled", "data-action-states": "[{\"action\":\"hover\",\"selector\":\".y-dropdown .ms-Dropdown\"},{\"action\":\"click\",\"selector\":\".y-dropdown .ms-Dropdown\"},{\"action\":\"keyPress\",\"key\":\"ArrowDown\"},{\"action\":\"keyPress\",\"key\":\"Enter\"}]"} }
const { GutterSize } = require('../Block');

const options = [
  { key: 'A', text: 'Option A' },
  { key: 'B', text: 'Option B' },
  { key: 'C', text: 'Option C' },
];

class ControlledDropdown extends React.Component {
  constructor() {
    super();

    this.state = {};

    this.handleChanged = this.handleChanged.bind(this);
  }

  render() {
    const { selectedKey } = this.state;

    return (
      <Dropdown
        options={options}
        placeHolder="Please select..."
        selectedKey={selectedKey}
        onChanged={this.handleChanged}
      />
    );
  }

  handleChanged(key) {
    action('dropdown changed')(key);
    this.setState({ selectedKey: key });
  }
}

<div style={{ paddingBottom: '100px' }}>
  <Block bottomSpacing={GutterSize.LARGE}>
    In this example, the state of the <Strong>Dropdown</Strong> is controlled by the parent component using the <Strong>selectedKey</Strong> and <Strong>onChanged</Strong> props.
  </Block>
  <ControlledDropdown />
</div>
```

Uncontrolled dropdown:

```js { "props": { "data-description": "uncontrolled" }, "data-action-states": "[]" }
const { GutterSize } = require('../Block');

const options = [
  { key: 'A', text: 'Option A' },
  { key: 'B', text: 'Option B' },
  { key: 'C', text: 'Option C' },
];

<div>
  <Block bottomSpacing={GutterSize.LARGE}>
    In this example, the state of the <Strong>Dropdown</Strong> is maintained within the component, rather than being set using <Strong>selectedKey</Strong>. This might be used when a parent component is listening to the <Strong>onChanged</Strong> event and doesn't need to manage the dropdown's visual state, but this is generally not recommended.
  </Block>
  <Dropdown
    placeHolder="Please select..."
    options={options}
    onChanged={action('dropdown changed')}
  />
</div>
```


With option templates, dividers and headers:

```js { "props": { "data-description": "with templates dividers headers", "data-action-states": "[{\"action\":\"hover\",\"selector\":\".y-dropdown .ms-Dropdown\"},{\"action\":\"click\",\"selector\":\".y-dropdown .ms-Dropdown\"},{\"action\":\"keyPress\",\"key\":\"ArrowDown\"},{\"action\":\"keyPress\",\"key\":\"Enter\"}]"} }
const { GutterSize } = require('../Block');
const { DropdownMenuItemType } = require('.');
const AttachIcon = require('../Icon/icons/Attach').default;
const GroupIcon = require('../Icon/icons/Group').default;
const FlagIcon = require('../Icon/icons/Flag').default;
const LockIcon = require('../Icon/icons/Lock').default;
const PhoneIcon = require('../Icon/icons/Phone').default;
const text = 'The quick brown fox jumped over the lazy dog.';

const options = [
  { key: 'section1', text: 'Icon, Label and Text', itemType: DropdownMenuItemType.Header},
  { key: 'A', icon: GroupIcon, label: 'Group', text: 'Option A' },
  { key: 'B', icon: AttachIcon, label: 'Attachment', text: `${text} ${text} ${text} ${text}` },
  { key: 'divider1', text: 'Section', itemType: DropdownMenuItemType.Divider},
  { key: 'section2', text: 'Label and Text', itemType: DropdownMenuItemType.Header},
  { key: 'D', label: 'Group', text: `${text} ${text} ${text} ${text}` },
  { key: 'divider2', text: 'Section', itemType: DropdownMenuItemType.Divider},
  { key: 'section3', text: 'Plain Text', itemType: DropdownMenuItemType.Header},
  { key: 'I', text: `${text} ${text} ${text} ${text}` },
];

<div style={{ paddingBottom: '350px' }}>
  <Block bottomSpacing={GutterSize.LARGE}>
    This <Strong>Dropdown</Strong> renders section headers and dividers, and uses icons and labels in options.
  </Block>
  <Dropdown
    label="Please choose from the following options:"
    placeHolder="Please select..."
    options={options}
    onChanged={action('dropdown changed')}
  />
</div>
```

In a grid:

```js { "props": { "data-description": "in grid" } }
const { GutterSize } = require('../Block')
const { FixedGridColumn, FixedGridRow } = require('../FixedGrid')
const action = (msg) => (...args) => { console.log(msg, ...args) };

const options = [
  { key: 'A', text: 'Option A' },
  { key: 'B', text: 'Option B' },
  { key: 'C', text: 'Option C' },
];

<div>
  <Block bottomSpacing={GutterSize.LARGE}>
    A <Strong>Dropdown</Strong> fills 100% of its container's width. To manage its width, nest it within a layout component.
  </Block>
  <FixedGridRow>
    <FixedGridColumn fixed={true} width={100}>
      <Dropdown
        placeHolder="Prefix"
        options={options}
        onChanged={action('dropdown changed')}
      />
    </FixedGridColumn>
    <FixedGridColumn>
      <Dropdown
        placeHolder="Suffix"
        options={options}
        onChanged={action('dropdown changed')}
      />
    </FixedGridColumn>
  </FixedGridRow>
</div>
```
