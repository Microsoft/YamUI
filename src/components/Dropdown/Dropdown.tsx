/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { Dropdown as FabricDropdown, IDropdownOption, DropdownMenuItemType } from 'office-ui-fabric-react/lib/Dropdown';
import { join } from '../../util/classNames';
import { IconSize } from '../Icon';
import ChevronDownMed from '../Icon/icons/ChevronDownMed';
import Block from '../Block';
import Text, { TextSize } from '../Text';
import { FixedGridRow, FixedGridColumn, GutterSize } from '../FixedGrid';
import { DropdownProps } from './Dropdown.types';
import './Dropdown.css';

/**
 * A `Dropdown` is a list in which the selected item is always visible, and the others are visible
 * on demand by clicking a button.
 */
export default class Dropdown extends React.Component<DropdownProps> {
  public render() {
    const { ariaLabel, className, label, placeHolder, selectedKey } = this.props;
    const calloutProps = {
      directionalHintFixed: false,
      doNotLayer: true,
      preventDismissOnScroll: true,
    };

    return (
      <div className={join(['y-dropdown', className])}>
        <FabricDropdown
          ariaLabel={ariaLabel}
          label={label}
          options={this.getFabricOptions()}
          placeHolder={placeHolder}
          selectedKey={selectedKey}
          onChanged={this.handleChanged}
          onRenderCaretDown={this.getChevronIcon}
          onRenderOption={this.getOption}
          onRenderTitle={this.getSelectedOption}
          onRenderPlaceHolder={this.getPlaceholder}
          calloutProps={calloutProps}
        />
      </div>
    );
  }

  private getChevronIcon = () => {
    return <ChevronDownMed className="y-dropdown__caretDown" size={IconSize.MEDIUM} />;
  };

  private getOptionText(label: string | undefined, text: string): React.ReactNode {
    if (label) {
      return (
        <span>
          <Text bold={true}>{label}:</Text> {text}
        </span>
      );
    }
    return <span>{text}</span>;
  }

  private getOptionContent(item: IDropdownOption) {
    const label = item.data && item.data.label;
    const Icon = item.data && item.data.icon;
    const iconNode = Icon && <Icon size={IconSize.MEDIUM} block={true} />;
    const textContent = this.getOptionText(label, item.text);

    if (iconNode) {
      return (
        <FixedGridRow gutterSize={GutterSize.SMALL}>
          <FixedGridColumn fixed={true} className="y-dropdown--option-icon">
            {iconNode}
          </FixedGridColumn>
          <FixedGridColumn className="y-dropdown--ellipsed-title">{textContent}</FixedGridColumn>
        </FixedGridRow>
      );
    }

    return <div className="y-dropdown--ellipsed-title">{textContent}</div>;
  }

  private getPlaceholder = () => {
    return <Block textSize={TextSize.MEDIUM_SUB}>{this.props.placeHolder}</Block>;
  };

  private getOption = (item?: IDropdownOption) => {
    if (!item) {
      return null;
    }

    const isHeader = item.itemType === DropdownMenuItemType.Header;

    if (isHeader) {
      return (
        <Block className="y-dropdown--item y-dropdown--header" textSize={TextSize.SMALL}>
          {item.text}
        </Block>
      );
    }

    return (
      <Block className="y-dropdown--item" textSize={TextSize.MEDIUM_SUB}>
        {this.getOptionContent(item)}
      </Block>
    );
  };

  // Fabric Dropdown can support multiple selection. Even though we're not supporting it, TypeScript
  // requires we handle these cases. So some branches cannot actually be reached/tested. RIP 100% coverage :(
  private getSelectedOption = (item?: IDropdownOption | IDropdownOption[]) => {
    if (Array.isArray(item)) {
      return item.length > 0 ? this.getOption(item[0]) : this.getOption();
    }
    return this.getOption(item);
  };

  private getFabricOptions() {
    return this.props.options.map(item => {
      return {
        key: item.key,
        text: item.text,
        itemType: item.itemType,
        data: {
          icon: item.icon,
          label: item.label,
        },
      };
    });
  }

  private handleChanged = (option: IDropdownOption) => {
    const { onChanged } = this.props;
    if (onChanged) {
      onChanged(option.key);
    }
  };
}
