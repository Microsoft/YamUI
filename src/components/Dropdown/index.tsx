/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { Dropdown as FabricDropdown, IDropdownOption } from 'office-ui-fabric-react/lib/Dropdown';
import { BaseComponentProps } from '../../util/BaseComponent/props';
import './dropdown.css';

export type DropdownOptionKey = string | number;
export type DropdownOption = IDropdownOption;

export interface DropdownProps extends BaseComponentProps {
  /**
   * Items to be presented to the user.
  **/
  options: DropdownOption[];

  /**
   * Label text that will be rendered above the dropdown.
  **/
  label?: string;

  /**
   * Use to specify an initially selected option.
  **/
  selectedKey?: DropdownOptionKey;

  /**
   * Function to call when user changes the selected item.
  **/
  onChanged?: (key: DropdownOptionKey) => void;

  /**
   * Input placeholder text. Displayed until option is selected.
  **/
  placeHolder?: string;
}

export default class Dropdown extends React.Component<DropdownProps, {}> {
  constructor() {
    super();
    this.handleChanged = this.handleChanged.bind(this);
  }

  render() {
    const { options, selectedKey, onChanged, placeHolder, label } = this.props;
    return (
      <div className={this.getClasses()}>
        <FabricDropdown
          options={options}
          selectedKey={selectedKey}
          onChanged={this.handleChanged}
          placeHolder={placeHolder}
          label={label}
        />
      </div>
    );
  }

  private getClasses() {
    const classes: string[] = [
      'y-dropdown',
    ];
    if (this.props.className) {
      classes.push(this.props.className);
    }
    return classes.join(' ');
  }

  private handleChanged(option: DropdownOption) {
    this.props.onChanged && this.props.onChanged(option.key);
  }
}
