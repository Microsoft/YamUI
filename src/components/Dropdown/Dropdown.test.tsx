/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import * as React from 'react';
import { mount, ReactWrapper, shallow, ShallowWrapper } from 'enzyme';
import {
  Dropdown as FabricDropdown,
  IDropdownProps,
} from 'office-ui-fabric-react/lib/components/Dropdown';
import Dropdown, { DropdownProps, DropdownMenuItemType } from '.';
import AddIcon from '../Icon/icons/Add';

describe('<Dropdown />', () => {
  let component: ShallowWrapper<DropdownProps, {}>;
  let fullComponent: ReactWrapper<DropdownProps, {}>;

  describe('with default options', () => {
    beforeEach(() => {
      component = shallow(<Dropdown options={[]} />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with additional className', () => {
    beforeEach(() => {
      component = shallow(<Dropdown options={[]} className="TEST_CLASSNAME" />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with all props', () => {
    beforeEach(() => {
      component = shallow(
        <Dropdown
          options={[{ key: 'A', text: 'Option a' }, { key: 'B', text: 'Option b' }]}
          className="TEST_CLASSNAME"
          selectedKey="A"
          placeHolder="TEST_PLACEHOLDER"
          label="TEST_LABEL"
          onChanged={jest.fn()}
        />,
      );
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('option templates', () => {
    describe('with icon', () => {
      beforeEach(() => {
        const options = [{ key: 'A', text: 'with icon', icon: AddIcon }];
        fullComponent = mount(<Dropdown options={options} />);
        fullComponent.find('.ms-Dropdown').simulate('click');
      });

      it('matches its snapshot', () => {
        expect(fullComponent.find('.ms-Dropdown-items .y-dropdown--item')).toMatchSnapshot();
      });
    });

    describe('with label', () => {
      beforeEach(() => {
        const options = [{ key: 'A', text: 'with icon', label: 'label' }];
        fullComponent = mount(<Dropdown options={options} />);
        fullComponent.find('.ms-Dropdown').simulate('click');
      });

      it('matches its snapshot', () => {
        expect(fullComponent.find('.ms-Dropdown-items .y-dropdown--item')).toMatchSnapshot();
      });
    });
  });

  describe('with divider', () => {
    beforeEach(() => {
      const options = [
        { key: 'divider', text: 'divider1', itemType: DropdownMenuItemType.Divider },
      ];
      fullComponent = mount(<Dropdown options={options} />);
      fullComponent.find('.ms-Dropdown').simulate('click');
    });

    it('matches its snapshot', () => {
      expect(fullComponent.find('.ms-Dropdown-items .y-dropdown--item')).toMatchSnapshot();
    });
  });

  describe('with section header', () => {
    beforeEach(() => {
      const options = [{ key: 'header', text: 'header1', itemType: DropdownMenuItemType.Header }];
      fullComponent = mount(<Dropdown options={options} />);
      fullComponent.find('.ms-Dropdown').simulate('click');
    });

    it('matches its snapshot', () => {
      expect(fullComponent.find('.ms-Dropdown-items .y-dropdown--item')).toMatchSnapshot();
    });
  });

  describe('when fully rendered', () => {
    beforeEach(() => {
      fullComponent = mount(<Dropdown options={[]} />);
    });

    it('displays our custom caret icon', () => {
      expect(fullComponent.find('.y-dropdown__caretDown').length).toBe(1);
    });
  });

  describe('with onChanged handler', () => {
    let callback: jest.Mock;
    const options = [{ key: 'A', text: 'Option a' }, { key: 'B', text: 'Option b' }];

    beforeEach(() => {
      callback = jest.fn();
      component = shallow(<Dropdown options={options} onChanged={callback} />);
    });

    describe('when an option is selected', () => {
      let fabricDropdown: ShallowWrapper<IDropdownProps, {}>;

      beforeEach(() => {
        fabricDropdown = component.find(FabricDropdown);
        fabricDropdown.simulate('changed', options[1]);
      });

      it('triggers the onChanged callback with key', () => {
        expect(callback).toHaveBeenCalledWith(options[1].key);
      });
    });
  });

  describe('when onRenderTitle is called with null', () => {
    let result: React.ReactNode | null;

    beforeEach(() => {
      const fabricDropdown = component.find(FabricDropdown);
      const onRenderTitle = fabricDropdown.prop('onRenderTitle');
      if (onRenderTitle) {
        result = onRenderTitle(undefined);
      }
    });

    it('returns the expected value', () => {
      expect(result).toMatchSnapshot();
    });
  });

  describe('when onRenderTitle is called with an empty array', () => {
    let result: React.ReactNode | null;

    beforeEach(() => {
      const fabricDropdown = component.find(FabricDropdown);
      const onRenderTitle = fabricDropdown.prop('onRenderTitle');
      if (onRenderTitle) {
        result = onRenderTitle([]);
      }
    });

    it('renders as expected', () => {
      expect(result).toMatchSnapshot();
    });
  });

  describe('when onRenderTitle is called with an array of items', () => {
    let result: React.ReactNode | null;

    beforeEach(() => {
      const fabricDropdown = component.find(FabricDropdown);
      const onRenderTitle = fabricDropdown.prop('onRenderTitle');
      if (onRenderTitle) {
        result = onRenderTitle([
          { itemType: DropdownMenuItemType.Header, key: '1', text: 'foo' },
          { key: '2', text: 'bar' },
        ]);
      }
    });

    it('renders as expected', () => {
      expect(result).toMatchSnapshot();
    });
  });
});
