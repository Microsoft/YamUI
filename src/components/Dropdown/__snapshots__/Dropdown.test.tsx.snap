// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> option templates with icon matches its snapshot 1`] = `
<div
  className=
      y-block
      y-textSize-mediumSub
      y-dropdown--item
      {
        font-size: 1.4rem;
        line-height: 2.0rem;
      }
      & .y-icon {
        height: 1.4rem;
        width: 1.4rem;
      }
      & .y-text__ellipsis {
        vertical-align: -0.5rem;
      }
>
  <div
    className=
        y-block--inner
        {
          overflow-wrap: break-word;
          word-wrap: break-word;
        }
  >
    <FixedGridRow
      gutterSize="small"
    >
      <div
        className=
            y-fixedGridRow
            {
              display: flex;
            }
            & > .y-fixedGridColumn {
              margin-left: 0.8rem;
            }
            & > .y-fixedGridColumn:first-child {
              margin-left: 0px;
            }
      >
        <FixedGridColumn
          className="y-dropdown--option-icon"
          fixed={true}
        >
          <div
            className=
                y-fixedGridColumn
                y-dropdown--option-icon
                {
                  flex-basis: auto;
                  flex-grow: 0;
                  flex-shrink: 1;
                  min-width: auto;
                }
          >
            <Add
              block={true}
              size="16"
            >
              <svg
                aria-hidden="true"
                className=
                    y-icon
                    {
                      display: block;
                      fill: currentColor;
                      pointer-events: none;
                      position: relative;
                      top: 0.2rem;
                      vertical-align: 0;
                    }
                focusable="false"
                style={
                  Object {
                    "height": "16px",
                    "width": "16px",
                  }
                }
                tabIndex={-1}
                viewBox="0 0 240 240"
              >
                <path
                  d="M240 112.5v15H127.5V240h-15V127.5H0v-15h112.5V0h15v112.5z"
                  fillRule="evenodd"
                />
              </svg>
            </Add>
          </div>
        </FixedGridColumn>
        <FixedGridColumn
          className="y-dropdown--ellipsed-title"
        >
          <div
            className=
                y-fixedGridColumn
                y-dropdown--ellipsed-title
                {
                  flex-basis: 100%;
                  flex-grow: 1;
                  flex-shrink: 1;
                  min-width: 0px;
                }
          >
            <span>
              with icon
            </span>
          </div>
        </FixedGridColumn>
      </div>
    </FixedGridRow>
  </div>
</div>
`;

exports[`<Dropdown /> option templates with label matches its snapshot 1`] = `
<div
  className=
      y-block
      y-textSize-mediumSub
      y-dropdown--item
      {
        font-size: 1.4rem;
        line-height: 2.0rem;
      }
      & .y-icon {
        height: 1.4rem;
        width: 1.4rem;
      }
      & .y-text__ellipsis {
        vertical-align: -0.5rem;
      }
>
  <div
    className=
        y-block--inner
        {
          overflow-wrap: break-word;
          word-wrap: break-word;
        }
  >
    <div
      className="y-dropdown--ellipsed-title"
    >
      <span>
        <CustomizedText
          bold={true}
        >
          <CustomizableText
            bold={true}
            theme={
              Object {
                "disableGlobalClassNames": false,
                "fonts": Object {
                  "large": Object {
                    "fontSize": "1.8rem",
                    "lineHeight": "2.4rem",
                  },
                  "medium": Object {
                    "fontSize": "1.5rem",
                    "lineHeight": "2.0rem",
                  },
                  "mediumPlus": Object {
                    "MozOsxFontSmoothing": "grayscale",
                    "WebkitFontSmoothing": "antialiased",
                    "fontFamily": "'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif",
                    "fontSize": "15px",
                    "fontWeight": 400,
                  },
                  "mega": Object {
                    "MozOsxFontSmoothing": "grayscale",
                    "WebkitFontSmoothing": "antialiased",
                    "fontFamily": "'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif",
                    "fontSize": "72px",
                    "fontWeight": 100,
                  },
                  "small": Object {
                    "fontSize": "1.2rem",
                    "lineHeight": "1.6rem",
                  },
                  "smallPlus": Object {
                    "fontSize": "1.4rem",
                    "lineHeight": "2.0rem",
                  },
                  "superLarge": Object {
                    "MozOsxFontSmoothing": "grayscale",
                    "WebkitFontSmoothing": "antialiased",
                    "fontFamily": "'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif",
                    "fontSize": "42px",
                    "fontWeight": 100,
                  },
                  "tiny": Object {
                    "MozOsxFontSmoothing": "grayscale",
                    "WebkitFontSmoothing": "antialiased",
                    "fontFamily": "'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif",
                    "fontSize": "10px",
                    "fontWeight": 600,
                  },
                  "xLarge": Object {
                    "fontSize": "2.2rem",
                    "lineHeight": "2.8rem",
                  },
                  "xSmall": Object {
                    "fontSize": "1.0rem",
                    "lineHeight": "1.6rem",
                  },
                  "xxLarge": Object {
                    "fontSize": "2.4rem",
                    "lineHeight": "2.8rem",
                  },
                },
                "isInverted": false,
                "palette": Object {
                  "accent": "#6c98d9",
                  "black": "#000000",
                  "blackTranslucent40": "rgba(0,0,0,.4)",
                  "blue": "#0078d4",
                  "blueDark": "#002050",
                  "blueLight": "#00bcf2",
                  "blueMid": "#00188f",
                  "green": "#107c10",
                  "greenDark": "#004b1c",
                  "greenLight": "#bad80a",
                  "magenta": "#b4009e",
                  "magentaDark": "#5c005c",
                  "magentaLight": "#e3008c",
                  "neutralDark": "#dde0e6",
                  "neutralLight": "#edeff2",
                  "neutralLighter": "#f3f5f8",
                  "neutralLighterAlt": "#f8f8f8",
                  "neutralPrimary": "#343A41",
                  "neutralPrimaryAlt": "#495361",
                  "neutralQuaternary": "#d0d0d0",
                  "neutralQuaternaryAlt": "#dadada",
                  "neutralSecondary": "#646d7a",
                  "neutralSecondaryAlt": "#767676",
                  "neutralTertiary": "#dde0e6",
                  "neutralTertiaryAlt": "#a8b0bd",
                  "orange": "#d83b01",
                  "orangeLight": "#ea4300",
                  "orangeLighter": "#ff8c00",
                  "purple": "#5c2d91",
                  "purpleDark": "#32145a",
                  "purpleLight": "#b4a0ff",
                  "red": "#d1423b",
                  "redDark": "#9e3028",
                  "teal": "#008272",
                  "tealDark": "#004b50",
                  "tealLight": "#00b294",
                  "themeDark": "#386cbb",
                  "themeDarkAlt": "#2f5c9f",
                  "themeDarker": "#264f8c",
                  "themeLight": "#c7e0f4",
                  "themeLighter": "#9ac3ff",
                  "themeLighterAlt": "#eff6fc",
                  "themePrimary": "#6c98d9",
                  "themeSecondary": "#2b88d8",
                  "themeTertiary": "#71afe5",
                  "white": "#ffffff",
                  "whiteTranslucent40": "rgba(255,255,255,.4)",
                  "yellow": "#ffc56c",
                  "yellowLight": "#fff100",
                },
                "semanticColors": Object {
                  "blockingBackground": "rgba(234, 67, 0, .2)",
                  "bodyBackground": "#ffffff",
                  "bodyDivider": "#dde0e6",
                  "bodyFrameBackground": "#ffffff",
                  "bodySubtext": "#646d7a",
                  "bodyText": "#343A41",
                  "bodyTextChecked": "#000000",
                  "buttonBackground": "#f3f5f8",
                  "buttonBackgroundChecked": "#a8b0bd",
                  "buttonBackgroundCheckedHovered": "#edeff2",
                  "buttonBackgroundHovered": "#edeff2",
                  "buttonBorder": "transparent",
                  "buttonText": "#495361",
                  "buttonTextChecked": "#dde0e6",
                  "buttonTextCheckedHovered": "#000000",
                  "buttonTextHovered": "#000000",
                  "disabledBackground": "#f3f5f8",
                  "disabledBodyText": "#a8b0bd",
                  "disabledSubtext": "#d0d0d0",
                  "disabledText": "#dde0e6",
                  "errorBackground": "#fb7f78",
                  "errorText": "#d1423b",
                  "focusBorder": "#000000",
                  "inputBackground": "#ffffff",
                  "inputBackgroundChecked": "#386cbb",
                  "inputBackgroundCheckedHovered": "#6c98d9",
                  "inputBorder": "#dde0e6",
                  "inputBorderHovered": "#dde0e6",
                  "inputFocusBorderAlt": "#6c98d9",
                  "inputForegroundChecked": "#ffffff",
                  "inputPlaceholderText": "#646d7a",
                  "link": "#386cbb",
                  "linkHovered": "#264f8c",
                  "listBackground": "#ffffff",
                  "listHeaderBackgroundHovered": "#f3f5f8",
                  "listHeaderBackgroundPressed": "#edeff2",
                  "listItemBackgroundChecked": "#edeff2",
                  "listItemBackgroundCheckedHovered": "#dadada",
                  "listItemBackgroundHovered": "#f3f5f8",
                  "listText": "#343A41",
                  "listTextColor": "#343A41",
                  "menuHeader": "#6c98d9",
                  "menuIcon": "#6c98d9",
                  "menuItemBackgroundChecked": "#edeff2",
                  "menuItemBackgroundHovered": "#f3f5f8",
                  "smallInputBorder": "#646d7a",
                  "successBackground": "rgba(186, 216, 10, .2)",
                  "warningBackground": "#ffc56c",
                  "warningHighlight": "#ffb900",
                  "warningText": "#333333",
                },
              }
            }
          >
            <span
              className=
                  y-text
                  {
                    display: inline-block;
                    font-weight: 600;
                  }
            >
              label
              :
            </span>
          </CustomizableText>
        </CustomizedText>
         
        with icon
      </span>
    </div>
  </div>
</div>
`;

exports[`<Dropdown /> when onRenderTitle is called with an array of items renders as expected 1`] = `
<CustomizedBlock
  className="y-dropdown--item y-dropdown--header"
  textSize="small"
>
  foo
</CustomizedBlock>
`;

exports[`<Dropdown /> when onRenderTitle is called with an empty array renders as expected 1`] = `null`;

exports[`<Dropdown /> when onRenderTitle is called with null returns the expected value 1`] = `null`;

exports[`<Dropdown /> with additional className matches its snapshot 1`] = `
<div
  className="y-dropdown TEST_CLASSNAME"
>
  <WithResponsiveMode
    calloutProps={
      Object {
        "directionalHintFixed": false,
        "doNotLayer": true,
        "preventDismissOnScroll": true,
      }
    }
    onChanged={[Function]}
    onRenderCaretDown={[Function]}
    onRenderOption={[Function]}
    onRenderPlaceHolder={[Function]}
    onRenderTitle={[Function]}
    options={Array []}
  />
</div>
`;

exports[`<Dropdown /> with all props matches its snapshot 1`] = `
<div
  className="y-dropdown TEST_CLASSNAME"
>
  <WithResponsiveMode
    ariaLabel="ARIA_LABEL"
    calloutProps={
      Object {
        "directionalHintFixed": false,
        "doNotLayer": true,
        "preventDismissOnScroll": true,
      }
    }
    label="TEST_LABEL"
    onChanged={[Function]}
    onRenderCaretDown={[Function]}
    onRenderOption={[Function]}
    onRenderPlaceHolder={[Function]}
    onRenderTitle={[Function]}
    options={
      Array [
        Object {
          "data": Object {
            "icon": undefined,
            "label": undefined,
          },
          "itemType": undefined,
          "key": "A",
          "text": "Option a",
        },
        Object {
          "data": Object {
            "icon": undefined,
            "label": undefined,
          },
          "itemType": undefined,
          "key": "B",
          "text": "Option b",
        },
      ]
    }
    placeHolder="TEST_PLACEHOLDER"
    selectedKey="A"
  />
</div>
`;

exports[`<Dropdown /> with default options matches its snapshot 1`] = `
<div
  className="y-dropdown"
>
  <WithResponsiveMode
    calloutProps={
      Object {
        "directionalHintFixed": false,
        "doNotLayer": true,
        "preventDismissOnScroll": true,
      }
    }
    onChanged={[Function]}
    onRenderCaretDown={[Function]}
    onRenderOption={[Function]}
    onRenderPlaceHolder={[Function]}
    onRenderTitle={[Function]}
    options={Array []}
  />
</div>
`;

exports[`<Dropdown /> with divider matches its snapshot 1`] = `null`;

exports[`<Dropdown /> with section header matches its snapshot 1`] = `
<div
  className=
      y-block
      y-textSize-small
      y-dropdown--item
      y-dropdown--header
      {
        font-size: 1.2rem;
        line-height: 1.6rem;
      }
      &.y-block .y-icon {
        top: 0.1rem;
      }
      & .y-icon {
        height: 1.2rem;
        width: 1.2rem;
      }
      & .y-text__ellipsis {
        vertical-align: -0.3rem;
      }
>
  <div
    className=
        y-block--inner
        {
          overflow-wrap: break-word;
          word-wrap: break-word;
        }
  >
    header1
  </div>
</div>
`;

exports[`<Dropdown /> without onChanged handler when an option is selected does not blow up 1`] = `
<div
  className="y-dropdown"
>
  <WithResponsiveMode
    calloutProps={
      Object {
        "directionalHintFixed": false,
        "doNotLayer": true,
        "preventDismissOnScroll": true,
      }
    }
    onChanged={[Function]}
    onRenderCaretDown={[Function]}
    onRenderOption={[Function]}
    onRenderPlaceHolder={[Function]}
    onRenderTitle={[Function]}
    options={
      Array [
        Object {
          "data": Object {
            "icon": undefined,
            "label": undefined,
          },
          "itemType": undefined,
          "key": "A",
          "text": "Option a",
        },
        Object {
          "data": Object {
            "icon": undefined,
            "label": undefined,
          },
          "itemType": undefined,
          "key": "B",
          "text": "Option b",
        },
      ]
    }
  />
</div>
`;
